src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,325 @@
+    BaseFloat samp_freq_input = input_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_input = input_matrix.NumCols(),  // number of samples in recording
+          num_input_channel = input_matrix.NumRows();  // number of channels in recording
+    KALDI_VLOG(1) << "sampling frequency of input: " << samp_freq_input
+                  << " #samples: " << num_samp_input
+                  << " #channel: " << num_input_channel;
+    KALDI_ASSERT(input_channel < num_input_channel);
+
+    WaveData rir_wave;
+    {
+      Input ki(rir_file);
+      rir_wave.Read(ki.Stream());
+    }
+    const Matrix<BaseFloat> &rir_matrix = rir_wave.Data();
+    BaseFloat samp_freq_rir = rir_wave.SampFreq();
+    int32 num_samp_rir = rir_matrix.NumCols(),
+          num_rir_channel = rir_matrix.NumRows();
+    KALDI_VLOG(1) << "sampling frequency of rir: " << samp_freq_rir
+                  << " #samples: " << num_samp_rir
+                  << " #channel: " << num_rir_channel;
+    if (!multi_channel_output) {
+      KALDI_ASSERT(rir_channel < num_rir_channel);
+    }
+
+    Matrix<BaseFloat> noise_matrix;
+    if (!noise_file.empty()) {
+      WaveData noise_wave;
+      {
+        Input ki(noise_file);
+        noise_wave.Read(ki.Stream());
+      }
+      noise_matrix = noise_wave.Data();
+      BaseFloat samp_freq_noise = noise_wave.SampFreq();
+      int32 num_samp_noise = noise_matrix.NumCols(),
+            num_noise_channel = noise_matrix.NumRows();
+      KALDI_VLOG(1) << "sampling frequency of noise: " << samp_freq_noise
+                    << " #samples: " << num_samp_noise
+                    << " #channel: " << num_noise_channel;
+      if (multi_channel_output) {
+        KALDI_ASSERT(num_rir_channel == num_noise_channel);
+      } else {
+        KALDI_ASSERT(noise_channel < num_noise_channel);
+      }
+    }
+
+    int32 num_output_channel = (multi_channel_output ? num_rir_channel : 1);
+    Matrix<BaseFloat> out_matrix(num_output_channel, num_samp_input);
+
+    for (int32 output_channel = 0; output_channel < num_output_channel; output_channel++) {
+      Vector<BaseFloat> input(num_samp_input);
+      input.CopyRowFromMat(input_matrix, input_channel);
+
+      int32 this_rir_channel = (multi_channel_output ? output_channel : rir_channel);
+      Vector<BaseFloat> rir(num_samp_rir);
+      rir.CopyRowFromMat(rir_matrix, this_rir_channel);
+      rir.Scale(1 / MaxAbsolute(rir));
+
+      Vector<BaseFloat> noise(0);
+      if (!noise_file.empty()) {
+        noise.Resize(noise_matrix.NumCols());
+        int32 this_noise_channel = (multi_channel_output ? output_channel : noise_channel);
+        noise.CopyRowFromMat(noise_matrix, this_noise_channel);
+      }
+
+      DoReverberation(rir, &noise, &input, samp_freq_rir, snr_db, test_mode);
+
+      float input_max = MaxAbsolute(input);
+      for (int32 j = 0; j < input.Dim(); j++)
+        input(j) = (0.99 * input(j) / input_max) * pow(2, 15);
```

Comment:
Guys, from this thread
https://sourceforge.net/p/kaldi/discussion/1355347/thread/4f07e084/?limit=25#1741
it looks like the 8-bit path of the code was used.
So I think the way forward right now could be to just modify the existing
8-bit and 32-bit code to scale the same as 16-bit.
This probably does make more sense than unnecessarily duplicating the code.

Dan


On Fri, Aug 21, 2015 at 7:39 PM, Daniel Povey <dpovey@gmail.com> wrote:

> I would be OK with simply dropping the 8 and 32-bit support and clearly
> documenting that the WaveReader supports only 16-bit PCM and that the range
> of the input/output is -2^15 to 2^15.  Probably we should also add support
> for truncating samples that are too large or small, but of course print a
> suitable warning message about truncated samples.
>
> Dan
>
>
> On Fri, Aug 21, 2015 at 5:52 PM, Kirill Katsnelson <
> notifications@github.com> wrote:
>
>> In src/featbin/wav-reverberate.cc
>> <https://github.com/kaldi-asr/kaldi/pull/76#discussion_r37689858>:
>>
>> > +      Vector<BaseFloat> rir(num_samp_rir);
>> > +      rir.CopyRowFromMat(rir_matrix, this_rir_channel);
>> > +      rir.Scale(1 / MaxAbsolute(rir));
>> > +
>> > +      Vector<BaseFloat> noise(0);
>> > +      if (!noise_file.empty()) {
>> > +        noise.Resize(noise_matrix.NumCols());
>> > +        int32 this_noise_channel = (multi_channel_output ? output_channel : noise_channel);
>> > +        noise.CopyRowFromMat(noise_matrix, this_noise_channel);
>> > +      }
>> > +
>> > +      DoReverberation(rir, &noise, &input, samp_freq_rir, snr_db, test_mode);
>> > +
>> > +      float input_max = MaxAbsolute(input);
>> > +      for (int32 j = 0; j < input.Dim(); j++)
>> > +        input(j) = (0.99 * input(j) / input_max) * pow(2, 15);
>>
>> fix the 8 and 32 bit cases to agree
>> Or drop them altogeter, because no one uses them anyway.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/76/files#r37689858>.
>>
>
>